src/services/schedule.ts

```diff
@@ -0,0 +1,136 @@
+import { PluginConfigId, PluginsServer } from '../types'
+import Piscina from 'piscina'
+import { processError } from '../error'
+import * as schedule from 'node-schedule'
+import Redlock from 'redlock'
+import * as Sentry from '@sentry/node'
+
+const lockedResource = 'plugin-server:locks:schedule'
+
+export async function startSchedule(
+    server: PluginsServer,
+    piscina: Piscina,
+    onLock?: () => void
+): Promise<() => Promise<void>> {
+    console.info(`⏰ Starting scheduling service`)
+
+    let stopped = false
+    let weHaveTheLock = false
+    let lock: Redlock.Lock
+    let lockTimeout: NodeJS.Timeout
+
+    const lockTTL = server.SCHEDULE_LOCK_TTL * 1000 // 60 sec
+    const retryDelay = lockTTL / 10 // 6 sec
+    const extendDelay = lockTTL / 2 // 30 sec
+
+    const redlock = new Redlock([server.redis], {
+        // we handle retires ourselves to have a way to cancel the promises on quit
+        // without this, the `await redlock.lock()` code will remain inflight and cause issues
+        retryCount: 0,
+    })
+
+    redlock.on('clientError', (error) => {
+        if (stopped) {
+            return
+        }
+        console.error('🔴 Redlock clientError', error)
+        Sentry.captureException(error)
+    })
+
+    const tryToGetTheLock = async () => {
+        try {
+            lock = await redlock.lock(lockedResource, lockTTL)
+            weHaveTheLock = true
+
+            console.info(`🔒 Scheduler lock acquired!`)
+
+            const extendLock = async () => {
+                if (stopped) {
+                    return
+                }
+                try {
+                    lock = await lock.extend(lockTTL)
+                    lockTimeout = setTimeout(extendLock, extendDelay)
+                } catch (error) {
+                    console.error('🔴 Redlock can not extend lock!', error)
+                    Sentry.captureException(error)
+                    weHaveTheLock = false
+                    lockTimeout = setTimeout(tryToGetTheLock, 0)
```

Comment:
Why the 0 ms timeouts? Also I think the realistic minimum is like 4 ms?

Reply:
I think this could actually be replaced by `process.nextTick`. The idea is to call this function as soon as possible, but outside the current stack. 4ms sounds quite random number to put in code though.

Reply:
This: https://stackoverflow.com/a/7221609 is pretty old but suggests that in Node there is no minimum like in browsers and in fact 0 is ran right away (though not _quite_, so maybe it does work the intended way here? I wouldn't totally bet on it though, `nextTick` sounds better for the job)

Reply:
Indeed, I want it to run right away, but async, so it's basically the same as [`process.nextTick`](https://nodejs.org/api/process.html#process_process_nexttick_callback_args). I chose `setTimeout` though just to be consistent with this `lockTimeout = ...` system, for the odd case it will get cancelled before it starts. Changed to `nextTick` now.

Reply:
Hm, still got `lockTimeout = setTimeout(tryToGetTheLock, 0)` though, do we want that `nextTick`ified as well or not?

Reply:
Let's go back to `setTimeout` actually https://dev.to/logicmason/settimeout-vs-setimmediate-vs-process-nexttick-3lj2